### PR TITLE
Glitchless 2: No Fun Allowed^2

### PR DESCRIPTION
**Pull request by chinese-soup**
- [x] Fixed quickgauss
- [x]  Fixed IHD
- [x]  "Fixed" NPCs going thru doors
- [x]  Fixed crash on hornet change level 
- [x] Fixed hitting buttons through walls
- [x] Fixed resetting weapon animations by saveloading
- [x] Fixed resetting vorts animations by saveloading
- [x] Fixed duplicating weapons by saveloading
- [x] Fixed https://github.com/ValveSoftware/halflife/issues/2458

### CODE REVIEW

File: dlls/gauss.cpp [hunk 1]
```diff
@@ -597,6 +602,7 @@ void CGauss::WeaponIdle( void )
 
 class CGaussAmmo : public CBasePlayerAmmo
 {
+
```
Comment (YaLTeR) [inline dlls/gauss.cpp]: Ah! My dear friend whom I almost started missing...

File: dlls/gauss.cpp [hunk 2]
```diff
@@ -209,7 +209,8 @@ void CGauss::SecondaryAttack()
 		SendWeaponAnim( GAUSS_SPINUP );
 		m_fInAttack = 1;
 		m_flTimeWeaponIdle = UTIL_WeaponTimeBase() + 0.5;
-		m_pPlayer->m_flStartCharge = gpGlobals->time;
+		//m_pPlayer->m_flStartCharge = gpGlobals->time;
```
Comment (YaLTeR) [inline dlls/gauss.cpp]: I think you can just drop the lines, no need to leave them commented out.

File: dlls/doors.cpp [hunk 3]
```diff
@@ -708,9 +708,11 @@ void CBaseDoor::Blocked( CBaseEntity *pOther )
 
 
 	// Hurt the blocker a little.
-	if ( pev->dmg )
-		pOther->TakeDamage( pev, pev, pev->dmg, DMG_CRUSH );
-
+	if (pev->dmg)
+	{
+		if(pev->dmg >= 0)
```
Comment (YaLTeR) [inline dlls/doors.cpp]: Just put the >= 0 in the original condition.

File: dlls/extdll.h [hunk 4]
```diff
@@ -34,6 +34,12 @@
 
 #include "archtypes.h"     // DAL
 
+
+// Misc C-runtime library headers
+#include "stdio.h"
+#include "stdlib.h"
+#include "math.h"
+
```
Comment (YaLTeR) [inline dlls/extdll.h]: This seems to belong to a different commit.

**Review by YaLTeR — CHANGES_REQUESTED** [review]

File: dlls/gauss.cpp (hunk 1 above)
Comment (chinese-soup) [reply dlls/gauss.cpp]: > draft

File: dlls/gauss.cpp (hunk 2 above)
Comment (chinese-soup) [reply dlls/gauss.cpp]: > draft

File: dlls/extdll.h (hunk 4 above)
Comment (chinese-soup) [reply dlls/extdll.h]: yep, linux compat or whatever, done

**Comment by chinese-soup** [comment]
Should be all ready, can't think of any more to fix that are "important" or I could consider a "glitch", map shit like skipping power up with tripmines is fine IMHO, code's _the shit_. I'm gonna test it next week with a test _run_ of whole game, I guess. image

Mostly scared fastfire fix might have f'd something up, all the others I've tested a lot and were all working fine.

**Known issue:** vort charge anim save-load reset fix is stoopid (MonsterThink+helper variable, since setting frame in Restore doesn't work), but it works so whatever. Also after reloading the already-charged-up-vort-beams disappear, just the correct frame is restored, not the charging effects themselves. (Vorts zap you at a certain AnimEvent defined in the model's sequence (24th frame for triggering the zap attack)). 

Saves from main game should all work even though save fields were changed, I think the engine just ignores the ones that are missing, but are Restored in the mod.

Feel free to code review. Don't forget to comment

File: dlls/doors.cpp [hunk 5]
```diff
@@ -708,9 +708,10 @@ void CBaseDoor::Blocked( CBaseEntity *pOther )
 
 
 	// Hurt the blocker a little.
-	if ( pev->dmg )
-		pOther->TakeDamage( pev, pev, pev->dmg, DMG_CRUSH );
-
+	if (pev->dmg && pev->dmg >= 0)
```
Comment (YaLTeR) [inline dlls/doors.cpp]:
```suggestion
	if (pev->dmg >= 0)
```

File: dlls/doors.cpp [hunk 6]
```diff
@@ -708,9 +708,10 @@ void CBaseDoor::Blocked( CBaseEntity *pOther )
...
+	{
```
Comment (YaLTeR) [inline dlls/doors.cpp]: No need to add {} I guess

File: dlls/cbase.h [hunk 7]
```diff
@@ -264,7 +264,13 @@ class CBaseEntity
 		return pEnt; 
 	}
 
-	static CBaseEntity *Instance( entvars_t *pev ) { return Instance( ENT( pev ) ); }
+	static CBaseEntity *Instance( entvars_t *pev )
+	{
+		// Fix hornet crash, among other things
+		if (!pev)
+			return Instance(ENT(0));
```
Comment (YaLTeR) [inline dlls/cbase.h]: How does this change anything though? Is ENT different based on type (int vs. pointer), if so then what does this return?

File: dlls/player.cpp [hunk 8]
```diff
@@ -117,6 +117,7 @@ TYPEDESCRIPTION	CBasePlayer::m_playerSaveData[] =
 	DEFINE_FIELD( CBasePlayer, m_pTank, FIELD_EHANDLE ),
 	DEFINE_FIELD( CBasePlayer, m_iHideHUD, FIELD_INTEGER ),
 	DEFINE_FIELD( CBasePlayer, m_iFOV, FIELD_INTEGER ),
+	DEFINE_FIELD( CBasePlayer, m_flNextAttack, FIELD_TIME ),
```
Comment (YaLTeR) [inline dlls/player.cpp:120]: Won't this also fix quickgauss without any changes in the quickgauss commit?

File: dlls/cbase.h (hunk 7 above)
Comment (chinese-soup) [reply dlls/cbase.h]: stolen from https://github.com/ValveSoftware/halflife/issues/3056#issuecomment-787112577

File: dlls/player.cpp (hunk 8 above)
Comment (chinese-soup) [reply dlls/player.cpp:120]: ~~yea that's what happens when you fix quickgauss first, then other shit~~, sry :))

File: dlls/doors.cpp (hunk 6 above)
Comment (chinese-soup) [reply dlls/doors.cpp]: aye, leftover from debug msg

File: dlls/cbase.h (hunk 7 above)
Comment (YaLTeR) [reply dlls/cbase.h]: I wonder if by returning the world the outside code will then be able to do any weird stuff to the worldspawn thereby causing new glitches

Comment (chinese-soup) [reply dlls/cbase.h]: guess i'll just do the pev->attacker !=nullptr blabla thing in chornet then

File: dlls/player.cpp (hunk 8 above)
Comment (chinese-soup) [reply dlls/player.cpp:120]: i'd still keep it (the "fix quickgauss" commit) since it's inline with the older HLSDK this way

https://github.com/a1batross/HLSDK_ancient/blob/master/DLLS/GAUSS.CPP

File: dlls/doors.cpp (hunk 5 above)
Comment (chinese-soup) [reply dlls/doors.cpp]: done (mostly posting, so i dont forget)

File: dlls/doors.cpp (hunk 6 above)
Comment (chinese-soup) [reply dlls/doors.cpp]: done

File: dlls/cbase.h (hunk 7 above)
Comment (chinese-soup) [reply dlls/cbase.h]: done.

File: README.md [hunk 9]
```diff
@@ -9,6 +9,18 @@ Half-Life no fun allowed edition
 - Fixed object boosting.
 - Fixed brute launching.
 
+New:
+
+- Fixed quickgauss
+- Fixed IHD
+- Fixed NPCs opening doors
+- Fixed crash on hornet change level
+- Fixed hitting buttons through walls
```
Comment (YaLTeR) [inline README.md]:
```suggestion
- Fixed pressing buttons through walls
```

File: README.md [hunk 10]
```diff
@@ -9,6 +9,18 @@ Half-Life no fun allowed edition
...
+- Fixed fastfire
+- Fixed vortigaunts charge being reset by saving and loading
```
Comment (YaLTeR) [inline README.md]:
```suggestion
- Fixed vortigaunts' charge being reset by saveloading
```

File: README.md [hunk 11]
```diff
@@ -9,6 +9,18 @@ Half-Life no fun allowed edition
...
+- Fixed duplicating weapons by saveloading
+- Fixed NPCs change yaw more slowly when game is running at higher FPS
```
Comment (YaLTeR) [inline README.md]:
```suggestion
- Fixed NPCs rotating slower when game is running at higher FPS
```

File: README.md [hunk 12]
```diff
@@ -9,6 +9,18 @@ Half-Life no fun allowed edition
 - Fixed object boosting.
 - Fixed brute launching.
 
+New:
+
+- Fixed quickgauss
```
Comment (YaLTeR) [inline README.md]: Add dots like the list above pls

File: dlls/player.cpp (hunk 8 above)
Comment (chinese-soup) [reply dlls/player.cpp:120]: Nah, you baited me. xd Fixed quickgauss commit has to be there. Just tested it as well just to be sure, but:

See matherunner's commentary here, that I've used as a base for the fix a year ago but already forgot: https://www.jwchong.com/hl/weapons.html#quickgauss

> Notice how the highlighted line is inexplicably commented out, so that m_flStartCharge does not get saved to the savestate. When the game is loaded from a save, this field will be zero. The calculation of current charging time is done by gpGlobals->time - m_pPlayer->m_flStartCharge throughout dlls/gauss.cpp, with gpGlobals->time being much larger than 4 in most cases.

That's why in older HLSDK it's not present, ye. D

**Comment by chinese-soup** [comment]
No worries, I broke the pressing buttons thru walls again at some point, gotta fix the fix.

**Comment by chinese-soup** [comment]
Ok yeah, let's ignore the player in the trace, that sounds better.

**Comment by chinese-soup** [comment]
Whops, forgot that you could've still pressed buttons through other entities like doors. ~~Fixed, final offer.~~

EDIT: Never mind, this way the Blast Pit buttons don't work unless you stand on them, cuz they tiny... The boys really did know why it was "UNDONE", didn't they...

**Comment by chinese-soup** [comment]
Ok, I am done with this, I am taking the L and make smol buttons (e.g. Blast Pit oxy & fuel) only pressable when aimed at directly (with crosshair). Could also do a flFraction > 0.98, or some shit, but that'd probably just introduce some spots where you could still press through walls...

I give up, it's a meme mod, anyway.

**Some examples**
https://user-images.githubusercontent.com/5108747/175794250-5a80a5ff-082e-4087-be0f-79af7e26567b.mp4

File: dlls/player.cpp [hunk 13]
```diff
@@ -1554,7 +1554,34 @@ void CBasePlayer::PlayerUse ( void )
...
+
+			TraceResult tr;
+			TraceResult tr2;
+
+			Vector vecSrc = EyePosition();
+			Vector vecEnd = VecBModelOrigin( pObject->pev );
+			UTIL_TraceLine( vecSrc, vecEnd, dont_ignore_monsters, ENT(pev), &tr ); // ENT(pev) is the player here
+
+			// Trace from crosshair
+			UTIL_TraceLine ( pev->origin + pev->view_ofs, pev->origin + pev->view_ofs + gpGlobals->v_forward * 128, dont_ignore_monsters, ENT(pev), &tr2);
```
Comment (YaLTeR) [inline dlls/player.cpp]: I guess this can be inside the if then?

File: dlls/player.cpp [hunk 14]
```diff
@@ -1554,7 +1554,34 @@ void CBasePlayer::PlayerUse ( void )
 			// This essentially moves the origin of the target to the corner nearest the player to test to see 
 			// if it's "hull" is in the view cone
 			vecLOS = UTIL_ClampVectorToBox( vecLOS, pObject->pev->size * 0.5 );
-			
+
+
```
Comment (YaLTeR) [inline dlls/player.cpp]: well if i'm requesting changes anyway might as well point out this double blank

File: dlls/player.cpp (hunk 13 above)
Comment (chinese-soup) [reply dlls/player.cpp]: truth

**Review by YaLTeR — APPROVED** [review]

**Comment by YaLTeR** [comment]
Thanks
